src/multiset.rs

```diff
@@ -0,0 +1,86 @@
+use rpds::RedBlackTreeMapSync;
+use std::hash::Hash;
+
+/// Immutable multiset implementation, which is threadsafe and hash stable, regardless of insertion order.
```

Comment:
I had a go at implementing this data structure in a separate file, keeping its implementation private (following some broad advice from @oflatt and @Alex-Fischman). Hopefully, that should make the code cleaner and make it easier to replace the implementation at a later date. I only exposed the operations needed in the sort so far.

Reply:
I think it would be better to keep the data structure in the same file as the sort, as that would let you keep everything private instead of `pub(crate)`.

Reply:
The only reason I wanted to put it in a separate file was so that the sort couldn't reach in and use any private methods. Or is that already the case even if I have it in the same file?

Reply:
I moved it into the same file, but in its own module to keep its implementation private from the outer module.